### PR TITLE
Fix #4798 - Add Fuel-related Methods from OpenStudio-Standard

### LINKED ISSUE #4798: Add Fuel-related Methods from OpenStudio-Standards
Enhancement Request
--------------
Implement guel-related methods from OpenStudio-Standards that have proven useful but which have not yet been implemented in C++.

## Detailed Description

Originally described in https://github.com/NREL/OpenStudio/issues/4792, this is enhancement request is just the fuel-related methods, so the work is broken in two distinct PRs.

## Possible Implementation

[Siz.ThermalZone.rb](https://github.com/NREL/openstudio-standards/blob/ba5168e63a87ac25d6092a61658f0741a0f8e2e8/lib/openstudio-standards/hvac_sizing/Siz.ThermalZone.rb)
```
OpenStudio::Model::ThermalZone
  
  heating_fuels()
  
  cooling_fuels()
```

[Siz.HeatingCoolingFuels.rb](https://github.com/NREL/openstudio-standards/blob/ba5168e63a87ac25d6092a61658f0741a0f8e2e8/lib/openstudio-standards/hvac_sizing/Siz.HeatingCoolingFuels.rb) 
```
OpenStudio::Model::Model

  plant_loop_heating_fuels()
  
  plant_loop_cooling_fuels()
  
  coil_heating_fuels()
  
  coil_heat_type()
  
  airloop_primary_heat_type()
  
  zone_equipment_heat_type()
  
  coil_cooling_fuels()
  
  zone_equipment_heating_fuels()
  
  zone_equipment_cooling_fuels()
  
  zone_airloop_heating_fuels()
  
  zone_airloop_cooling_fuels()
```

Also add all the Ruby methods defined in these files that aren't already in OpenStudio C++, which will enable a lot of code and files to be removed from openstudio-results measure, as pointed out by @eringold.

https://github.com/NREL/openstudio-common-measures-gem/tree/develop/lib/measures/openstudio_results/resources

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #4798

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/model/AirLoopHVACOutdoorAirSystem.cpp [hunk 1]
```diff
@@ -612,6 +615,69 @@ namespace model {
...
+
+    std::vector<FuelType> AirLoopHVACOutdoorAirSystem_Impl::coolingFuelTypes() const {
+      std::set<FuelType> result;
+      for (const auto& comp : subsetCastVector<HVACComponent>(components())) {
+        for (auto& ft : comp.coolingFuelTypes()) {
+          result.insert(ft);
+        }
+      }
+      return {result.begin(), result.end()};
+    }
```
Comment (jmarrec) [inline src/model/AirLoopHVACOutdoorAirSystem.cpp:659]: An AirLoopHVACOutdoorAirSystem is listed as a component on an AirLoopHVAC. So we need to implement the logic for AirLoopHVACOutdoorAirSystem as well.
Currently I'm checking each `components`, meaning regardless of whether it's on the **relief** stream or the **oa intake** stream. Does that sound right to you @asparke2

Comment (jmarrec) [reply src/model/AirLoopHVACOutdoorAirSystem.cpp:659]: Note to self: the HX:AirToAir:SensibleAndLatent appears **twice** in `components` (because it's on both stream). Even if we made it be something than no-op, that's fine, we're working with std::set

File: src/model/CoilHeatingWaterToAirHeatPumpEquationFit.cpp [hunk 2]
```diff
@@ -383,6 +386,32 @@ namespace model {
...
+    }
+
+    std::vector<AppGFuelType> CoilHeatingWaterToAirHeatPumpEquationFit_Impl::appGHeatingFuelTypes() const {
+      std::set<AppGFuelType> result;
+      result.insert(AppGFuelType::HeatPump);
+      if (auto p_ = plantLoop()) {
+        return p_->appGHeatingFuelTypes();
+      }
+      return {};
+    }
```
Comment (jmarrec) [inline src/model/CoilHeatingWaterToAirHeatPumpEquationFit.cpp:413]: I'm not 100% sure I did the WaterToAir / AirToWater heat pump coils correctly to be honest, a review would be nice.

File: src/model/WaterHeaterHeatPump.cpp [hunk 3]
```diff
@@ -544,6 +546,22 @@ namespace model {
...
+      return {};
+    }
+
+    std::vector<FuelType> WaterHeaterHeatPump_Impl::heatingFuelTypes() const {
+      return {FuelType::Electricity};
+    }
+
+    std::vector<AppGFuelType> WaterHeaterHeatPump_Impl::appGHeatingFuelTypes() const {
+      return {AppGFuelType::HeatPump};  // TODO: openstudio-standards uses Electric
+    }
```
Comment (jmarrec) [inline src/model/WaterHeaterHeatPump.cpp:563]: This is most likely wrong.

* Do we want an HPWH in a zone affecting the answer for that zone's?
* It should probably go check what it's dxCoil is actually doing no?

File: src/model/ThermalZone.cpp [hunk 4]
```diff
@@ -2631,6 +2627,144 @@ namespace model {
...
+
+    std::vector<FuelType> ThermalZone_Impl::coolingFuelTypes() const {
+      std::set<FuelType> result;
+      for (const auto& comp : subsetCastVector<HVACComponent>(equipment())) {
+        for (auto& ft : comp.coolingFuelTypes()) {
+          result.insert(ft);
+        }
+      }
+      return {result.begin(), result.end()};
+    }
```
Comment (jmarrec) [inline src/model/ThermalZone.cpp:2746]: ThermalZone checks each individual equipment.

File: src/model/HVACComponent_Impl.hpp [hunk 5]
```diff
@@ -114,6 +118,11 @@ namespace model {
 
       virtual std::vector<std::string> emsInternalVariableNames() const override;
 
+      virtual ComponentType componentType() const = 0;
+      virtual std::vector<FuelType> coolingFuelTypes() const = 0;
+      virtual std::vector<FuelType> heatingFuelTypes() const = 0;
+      virtual std::vector<AppGFuelType> appGHeatingFuelTypes() const = 0;
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/HVACComponent_Impl.hpp:124]: As you can see in the implementation, I'm often using a `std::set` internally so that the returned `std::vector` is actually unique (and sorted). 
I didn't go with `std::set` because I was unsure how SWIG would play with a `std::set`...

Would you prefer `std::set? @kbenne @tijcolem @asparke2 please?

File: src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.hpp [hunk 6]
```diff
@@ -55,7 +58,7 @@ namespace energyplus {
     NONE
   };
 
-  ENERGYPLUS_API ComponentType componentType(const openstudio::model::HVACComponent& component);
+  OS_DEPRECATED ENERGYPLUS_API openstudio::energyplus::ComponentType componentType(const openstudio::model::HVACComponent& component);
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.hpp:61]: Deprecating the existing implementation that was in E+ FT.

File: src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.hpp [hunk 7]
```diff
@@ -75,7 +78,7 @@ namespace energyplus {
    * * If there is no cooling, no heating, no "both": None
    * * All other cases: "both"
    */
-  ENERGYPLUS_API ComponentType plantLoopType(const openstudio::model::PlantLoop& plantLoop);
+  OS_DEPRECATED ENERGYPLUS_API openstudio::energyplus::ComponentType plantLoopType(const openstudio::model::PlantLoop& plantLoop);
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.hpp:81]: Same here. I chose to use Loop_Impl::componentType to match HVACComponent_Impl, and note Loop_Impl::loopType. Maybe that was a dumb idea, don't know

File: src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.cpp [hunk 8]
```diff
@@ -158,6 +159,28 @@ namespace energyplus {
...
+
+  openstudio::energyplus::ComponentType plantLoopType(const PlantLoop& plantLoop) {
+    LOG_FREE(Warn, "openstudio.energyplus.plantLoopType", "Method is deprecated. Use Loop::componentType instead");
+    return convertOSEnumToEnum(plantLoop.componentType());
+  }
+
+  openstudio::energyplus::ComponentType componentType(const openstudio::model::HVACComponent& component) {
+    LOG_FREE(Warn, "openstudio.energyplus.componentType", "Method is deprecated. Use HVACComponent::componentType instead");
+    return convertOSEnumToEnum(component.componentType());
+  }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.cpp:182]: Deprecated functions in FT now call the model API. I have to do enum shenanigans since the FT had an `enum class` and now we use an OPENSTUDIO_ENUM (which is a class)

File: src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp [hunk 9]
```diff
@@ -274,6 +273,22 @@ namespace model {
...
+      return {};
+    }
+
+    std::vector<FuelType> SolarCollectorFlatPlatePhotovoltaicThermal_Impl::heatingFuelTypes() const {
+      return {FuelType::EnergyTransfer};
+    }
+
+    std::vector<AppGFuelType> SolarCollectorFlatPlatePhotovoltaicThermal_Impl::appGHeatingFuelTypes() const {
+      return {AppGFuelType::Solar};
+    }
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp:290]: We use 'SolarEnergy' in openstudio-standards

File: src/utilities/data/DataEnums.hpp [hunk 10]
```diff
@@ -249,8 +249,72 @@ OPENSTUDIO_ENUM(BuildingSector,
...
+OPENSTUDIO_ENUM(ComponentType,
+  ((Heating))
+  ((Cooling))
+  ((Both))
+  ((None))
+);
+
 // clang-format on
 
+inline UTILITIES_API AppGFuelType convertFuelTypeToAppG(FuelType fuelType) {
```
Comment (mdahlhausen) [inline src/utilities/data/DataEnums.hpp:300]: Seems like this method will never return the AppG Fuel Types 'HeatPump', 'Solar', or 'Mixed' from the data enums above

File: src/model/SetpointManager.cpp [hunk 11]
```diff
@@ -177,6 +179,22 @@ namespace model {
       return boost::none;
     }
 
+    ComponentType SetpointManager_Impl::componentType() const {
+      return ComponentType::None;
+    }
+
+    std::vector<FuelType> SetpointManager_Impl::coolingFuelTypes() const {
+      return {FuelType::Electricity};
```
Comment (mdahlhausen) [inline src/model/SetpointManager.cpp:187]: why does a setpoint manager have a cooling fuel type?

File: src/model/SwimmingPoolIndoor.cpp [hunk 12]
```diff
@@ -377,6 +375,22 @@ namespace model {
       return result;
     }
 
+    ComponentType SwimmingPoolIndoor_Impl::componentType() const {
+      return ComponentType::None;
+    }
+
+    std::vector<FuelType> SwimmingPoolIndoor_Impl::coolingFuelTypes() const {
+      return {FuelType::Electricity};
```
Comment (mdahlhausen) [inline src/model/SwimmingPoolIndoor.cpp:383]: Why does a Swimming Pool have a cooling fuel type?

File: src/model/HumidifierSteamGas.cpp [hunk 13]
```diff
@@ -343,6 +344,22 @@ namespace model {
...
+    ComponentType HumidifierSteamGas_Impl::componentType() const {
+      return ComponentType::Heating;  // TODO: is that right?
+    }
+
+    std::vector<FuelType> HumidifierSteamGas_Impl::coolingFuelTypes() const {
+      return {};
+    }
+
+    std::vector<FuelType> HumidifierSteamGas_Impl::heatingFuelTypes() const {
+      return {FuelType::Gas};  // TODO: is that right?
```
Comment (mdahlhausen) [inline src/model/HumidifierSteamGas.cpp:356]: I'd say it's pretty clearly a natural gas device so you can class it as heating with fuel type gas / fuel

File: src/model/AirTerminalSingleDuctConstantVolumeReheat.cpp [hunk 14]
```diff
@@ -525,6 +525,52 @@ namespace model {
...
+        result.insert(ft);
+      }
+      if (auto a_ = airLoopHVAC()) {
+        for (auto ft : a_->appGHeatingFuelTypes()) {
+          result.insert(ft);
+        }
+      }
+      return {result.begin(), result.end()};
+    }
+
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply src/model/AirTerminalSingleDuctConstantVolumeReheat.cpp:573]: yeah, that's appropriate

File: src/model/ChillerElectricEIR.cpp [hunk 15]
```diff
@@ -817,6 +816,22 @@ namespace model {
       OS_ASSERT(result);
     }
 
+    ComponentType ChillerElectricEIR_Impl::componentType() const {
+      return ComponentType::Cooling;
+    }
+
+    std::vector<FuelType> ChillerElectricEIR_Impl::coolingFuelTypes() const {
+      return {FuelType::Electricity};
+    }
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply src/model/ChillerElectricEIR.cpp:825]: I think checking the condenser loops is consistent. It'd be a rare but possible condition where it wouldn't be electric.

File: src/model/CoolingTowerVariableSpeed.cpp [hunk 16]
```diff
@@ -748,6 +749,22 @@ namespace model {
...
+    std::vector<FuelType> CoolingTowerVariableSpeed_Impl::coolingFuelTypes() const {
+      return {FuelType::Electricity};
+    }
+
+    std::vector<FuelType> CoolingTowerVariableSpeed_Impl::heatingFuelTypes() const {
+      return {};
+    }
+
+    std::vector<AppGFuelType> CoolingTowerVariableSpeed_Impl::appGHeatingFuelTypes() const {
+      return {AppGFuelType::Electric};
```
Comment (mdahlhausen) [inline src/model/CoolingTowerVariableSpeed.cpp:765]: The other cooling tower objects don't have AppGFuelType::Electric specified in the return.  Either way is fine; just think they should be consistent.

File: src/model/ZoneHVACEnergyRecoveryVentilator.cpp [hunk 17]
```diff
@@ -363,6 +364,22 @@ namespace model {
...
+      return {};
+    }
+
+    std::vector<FuelType> ZoneHVACEnergyRecoveryVentilator_Impl::heatingFuelTypes() const {
+      return {};
+    }
+
+    std::vector<AppGFuelType> ZoneHVACEnergyRecoveryVentilator_Impl::appGHeatingFuelTypes() const {
+      return {};
+    }
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply src/model/ZoneHVACEnergyRecoveryVentilator.cpp:381]: Probably both or none.  Leaning towards none.

File: src/model/WaterHeaterHeatPump.cpp (hunk 3 above)
Comment (mdahlhausen) [reply src/model/WaterHeaterHeatPump.cpp:563]: I'd consider it more of a beneficial load than an HVAC component, so no.

File: src/model/AirLoopHVACUnitarySystem.cpp [hunk 18]
```diff
@@ -1327,6 +1325,71 @@ namespace model {
...
+      }
+      // TODO: is supplemental needed?
+      if (auto supHC_ = supplementalHeatingCoil()) {
+        for (auto ft : supHC_->heatingFuelTypes()) {
+          result.insert(ft);
+        }
+      }
+
+      return {result.begin(), result.end()};
+    }
```
[the remark this answers is not in the document]
Comment (mdahlhausen) [reply src/model/AirLoopHVACUnitarySystem.cpp:1375]: I think the supplemental heating coil is important, especially as some of the heatpump equipment has minimum operating temperatures below which the supplemental coil provides all of the heat.

File: src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (asparke2) [reply src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp:290]: Agree. SolarEnergy sounds good to me.

File: src/model/HVACComponent_Impl.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (asparke2) [reply src/model/HVACComponent_Impl.hpp:124]: I don't know enough C++ to comment on the implementation sidee, but on the return type you get from the API I'd prefer a std::vector<Whatever>.

File: src/model/GroundHeatExchangerVertical.cpp [hunk 19]
```diff
@@ -418,6 +419,22 @@ namespace model {
...
+      return {FuelType::EnergyTransfer};
+    }
+
+    std::vector<FuelType> GroundHeatExchangerVertical_Impl::heatingFuelTypes() const {
+      return {FuelType::EnergyTransfer};
+    }
+
+    std::vector<AppGFuelType> GroundHeatExchangerVertical_Impl::appGHeatingFuelTypes() const {
+      return {AppGFuelType::Other};
+    }
```
[the remark this answers is not in the document]
Comment (asparke2) [reply src/model/GroundHeatExchangerVertical.cpp:436]: Hmm. I think in this case how about GeoThermalEnergy? Then people can classify however they want.

File: src/model/WaterHeaterHeatPump.cpp (hunk 3 above)
Comment (asparke2) [reply src/model/WaterHeaterHeatPump.cpp:563]: Agree, a HPWH in a zone should just be considered a zone load, not a zone heating/cooling piece of HVAC equipment. If a HPWH is on a water loop that serves equipment that is heating a zone, then it's heating and cooling fuel type should be "Electricity."

File: src/model/SwimmingPoolIndoor.cpp (hunk 12 above)
Comment (asparke2) [reply src/model/SwimmingPoolIndoor.cpp:383]: agree, I would consider a pool a zone load, not a piece of intentional conditioning equipment.

File: src/model/CoolingTowerVariableSpeed.cpp (hunk 16 above)
Comment (asparke2) [reply src/model/CoolingTowerVariableSpeed.cpp:765]: Yeah, cooling towers should be Electric

**Review by asparke2 — APPROVED** [review]
@jmarrec as far as I can tell this looks good. My C++ is very rusty so I don't understand the implementation details around the forward translator. Conceptually whatever you did to make these methods always forced to be implemented sounds like a great idea because otherwise people will forget. Testing this is difficult, when we switch openstudio-standards over to use this code we may find an issue, but I don't see one now and I can't think of good tests unless you want to add some more detailed OSMs to the test suite. If there is an issue I suspect it'll be a corner case.

File: src/model/HumidifierSteamGas.cpp (hunk 13 above)
Comment (jmarrec) [reply src/model/HumidifierSteamGas.cpp:356]: My doubt is more about whether a Humidifier should be considered a heating element.

File: src/model/SetpointManager.cpp (hunk 11 above)
Comment (jmarrec) [reply src/model/SetpointManager.cpp:187]: That would be a problem between the keyboard and the chair. I probably copy pasted, shame on me!

File: src/utilities/data/DataEnums.hpp (hunk 10 above)
Comment (jmarrec) [reply src/utilities/data/DataEnums.hpp:300]: That's actually normal. The idea there is that any IDD field that takes a `\choice` with a list of fuel types could be automatically convert for convenience.

Solar for eg would be directly harcoded where appropriate

https://github.com/NREL/OpenStudio/blob/6b567a46a6898303630c668dcc3ba17775feb30e/src/model/SolarCollectorFlatPlateWater.cpp#L212-L214

File: src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp:290]: Adding new FuelType enum values makes me uneasy. Here's the code from develop: https://github.com/NREL/OpenStudio/blob/581522ca8e29fbae99b4680b3ee93ae2c595696b/src/utilities/data/DataEnums.hpp#L86-L102

This enum was meant to mimic a `\choice` "Fuel Type" field in the IDD, or anything that relates to Output:Meter / the E+ output reports, eg this

```
OS:Boiler:HotWater,
[...]
  A3, \field Fuel Type
       \type choice
       \default NaturalGas
       \key Electricity
       \key NaturalGas
       \key Propane
       \key FuelOilNo1
       \key FuelOilNo2
       \key Coal
       \key Diesel
       \key Gasoline
       \key OtherFuel1
       \key OtherFuel2
```

**Comment by jmarrec** [comment]
I wrote a jupyter notebook to parse the current methods I added a found  more than a few mistakes I had made. I also uploaded the result of that parsing as a table on https://docs.google.com/spreadsheets/d/1n6qUtJhIsMa8uS8VBQCik-62FJJPFMrGU6bO5L962P4/edit?usp=sharing . 
This should be helpful to spot check the current implementation

Note: It's not impossible that the parsing messes up a couple of cases and differs from the actual implementation (I used a ton of regexes...). But it should be a good starting point.

File: src/model/CoolingTowerVariableSpeed.cpp (hunk 16 above)
Comment (jmarrec) [reply src/model/CoolingTowerVariableSpeed.cpp:765]: Done in 0a4c4324aaa10372d723b6cdf685a8883481092c

File: src/model/WaterHeaterHeatPump.cpp (hunk 3 above)
Comment (jmarrec) [reply src/model/WaterHeaterHeatPump.cpp:563]: Hum. So ThermalZone will call all of its equipment and check the corresponding method...
PlantLoop would call of its supply side equipment and check the corresponding method...


A WaterHeaterHeatPump has a dXCoil and an WaterHeaterMixed. 

The WaterHeaterHeatPump is listed on the ThermalZone's equipment(). 

The PlantLoop lists the tank. The Tank goes to check the WaterHeaterHeatPump (+ its internal heater + any secondaryPlantLoop)

```ruby
m =  Model.new
hpwh= WaterHeaterHeatPump.new(m)
tank = hpwh.tank.to_WaterHeaterMixed.get
z = ThermalZone.new(m)
p = PlantLoop.new(m)
hpwh.addToThermalZone(z)
p.addSupplyBranchForComponent(tank)


tank.setHeaterMaximumCapacity(845000.0)
tank.setHeaterFuelType("Propane")
# This returns the Electricity from the WaterHeaterHeatPump (containingZoneHVACComponent)
tank.heatingFuelTypes.map(&:valueName)
=> ["Electricity", "Propane"]
tank.setHeaterMaximumCapacity(0.0)
=> true
# Cap is zero, so it only returns the WaterHeaterHeatPump
tank.heatingFuelTypes.map(&:valueName)
=> ["Electricity"]

# But thermal zones also checks the WaterHeaterHeatPump
z.heatingFuelTypes.map(&:valueName)
=> ["Electricity"]
```


How do we deal with this?

Comment (mdahlhausen) [reply src/model/WaterHeaterHeatPump.cpp:563]: Water Heater Heat Pumps and their tanks should return empty as a heating or cooling fuel; they are not intended to provide space heating.

Comment (jmarrec) [reply src/model/WaterHeaterHeatPump.cpp:563]: Well, the HPWH's tank can be placed on the supply side of a PlantLoop, in which case it's heating. So I suppose the tank should check for that. The WaterHeaterHeatPump itself could return empty, so the ThermalZone's methods aren't affected. Sound good?

File: src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/SolarCollectorFlatPlatePhotovoltaicThermal.cpp:308]: Well, I suppose the EnergyTransfer is not a Fuel Type either, and probably comes from the "Resource Type" field of the an EMS/PythonPlugin OutputVariable.

```
EnergyManagementSystem:MeteredOutputVariable,
   [...]
   A6, \field Resource Type
       \note This field is optional for regular output variables with "Type of Data in Variable" set to
       \note either Averaged or Summed. For Metered variables, this field is required.
       \note Choose the type of fuel, water, electricity, pollution or heat rate that should be metered.
       \type choice
       \key Electricity
       \key NaturalGas
       \key Gasoline
       \key Diesel
       \key Coal
       \key FuelOilNo1
       \key FuelOilNo2
       \key Propane
       \key OtherFuel1
       \key OtherFuel2
       \key WaterUse
       \key OnSiteWaterProduced
       \key MainsWaterSupply
       \key RainWaterCollected
       \key WellWaterDrawn
       \key CondensateWaterCollected
       \key EnergyTransfer
       \key Steam
       \key DistrictCooling
       \key DistrictHeating
       \key ElectricityProducedOnSite
       \key SolarWaterHeating
       \key SolarAirHeating
```

File: src/model/ChillerElectricEIR.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/ChillerElectricEIR.cpp]: Done in 37e1ae80c8b20505888aa732595c1325d813a80c

File: src/model/ZoneHVACTerminalUnitVariableRefrigerantFlow.cpp [hunk 20]
```diff
@@ -810,6 +810,77 @@ namespace model {
...
+      std::set<FuelType> result;
+      if (auto cc_ = coolingCoil()) {
+        for (auto ft : cc_->coolingFuelTypes()) {
+          result.insert(ft);
+        }
+      }
+
+      // TODO: check the AirConditionnerVariableRefrigerantFlow ( / FluidCtrl) that serves it too?
+      return {result.begin(), result.end()};
+    }
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/ZoneHVACTerminalUnitVariableRefrigerantFlow.cpp]: Done in 66226e4bdd8d237d584094a3c270c71dfcc9117f

File: src/model/CoilHeatingGasMultiStage.cpp [hunk 21]
```diff
@@ -351,6 +351,22 @@ namespace model {
...
+      return {};
+    }
+
+    std::vector<FuelType> CoilHeatingGasMultiStage_Impl::heatingFuelTypes() const {
+      return {FuelType::Gas};
+    }
+
+    std::vector<AppGFuelType> CoilHeatingGasMultiStage_Impl::appGHeatingFuelTypes() const {
+      return {AppGFuelType::Fuel};
+    }
```
Comment (jmarrec) [inline src/model/CoilHeatingGasMultiStage.cpp:368]: This one is Gas only

File: src/model/CoilHeatingWaterToAirHeatPumpEquationFit.cpp [hunk 22]
```diff
@@ -386,6 +386,36 @@ namespace model {
...
+    std::vector<AppGFuelType> CoilHeatingWaterToAirHeatPumpEquationFit_Impl::appGHeatingFuelTypes() const {
+      std::set<AppGFuelType> result;
+      result.insert(AppGFuelType::HeatPump);
+      if (auto p_ = plantLoop()) {
+        for (auto ft : p_->appGHeatingFuelTypes()) {
+          result.insert(ft);
+        }
+      }
+      return {result.begin(), result.end()};
+    }
```
Comment (jmarrec) [inline src/model/CoilHeatingWaterToAirHeatPumpEquationFit.cpp:417]: **Water** To **Air** coils: Electric/HeatPump + the PlantLoop that serves it.

File: src/model/CoilWaterHeatingAirToWaterHeatPumpWrapped.cpp [hunk 23]
```diff
@@ -326,6 +326,22 @@ namespace model {
...
+    //   return {};
+    // }
+
+    // std::vector<FuelType> CoilWaterHeatingAirToWaterHeatPumpWrapped_Impl::heatingFuelTypes() const {
+    //   return {FuelType::Electricity};
+    // }
+
+    // std::vector<AppGFuelType> CoilWaterHeatingAirToWaterHeatPumpWrapped_Impl::appGHeatingFuelTypes() const {
+    //   return {AppGFuelType::HeatPump};  // TODO: openstudio-standards uses Electric
+    // }
```
Comment (jmarrec) [inline src/model/CoilWaterHeatingAirToWaterHeatPumpWrapped.cpp:343]: CoilWaterHeatingAirToWaterHeatPumpWrapped is **NOT** an HVACComponent... That's a bit problematic...

File: src/model/ThermalStorageChilledWaterStratified.cpp [hunk 24]
```diff
@@ -742,6 +742,25 @@ namespace model {
...
+    ComponentType ThermalStorageChilledWaterStratified_Impl::componentType() const {
+      return ComponentType::Cooling;
+    }
+
+    std::vector<FuelType> ThermalStorageChilledWaterStratified_Impl::coolingFuelTypes() const {
+      if (auto p_ = secondaryPlantLoop()) {
+        return p_->coolingFuelTypes();
+      }
+      return {};
+    }
```
Comment (jmarrec) [inline src/model/ThermalStorageChilledWaterStratified.cpp:754]: * [ ] TODO: should componentType be either 'None' or 'Both' instead? Should coolingFuelTypes be empty?

cf ThermalStorageIceDetailed too.

File: src/model/ThermalStorageIceDetailed.cpp [hunk 25]
```diff
@@ -289,6 +291,22 @@ namespace model {
       return result;
     }
 
+    ComponentType ThermalStorageIceDetailed_Impl::componentType() const {
+      return ComponentType::Both;
+    }
+
+    std::vector<FuelType> ThermalStorageIceDetailed_Impl::coolingFuelTypes() const {
+      return {};
+    }
```
Comment (jmarrec) [inline src/model/ThermalStorageIceDetailed.cpp:300]: TODO

File: src/model/WaterHeaterHeatPump.cpp [hunk 26]
```diff
@@ -544,6 +546,25 @@ namespace model {
...
+      return {};
+    }
+
+    std::vector<FuelType> WaterHeaterHeatPump_Impl::heatingFuelTypes() const {
+      return {};
+    }
+
+    std::vector<AppGFuelType> WaterHeaterHeatPump_Impl::appGHeatingFuelTypes() const {
+      return {};
+    }
```
Comment (jmarrec) [inline src/model/WaterHeaterHeatPump.cpp:566]: cf HPWH

File: src/model/WaterHeaterMixed.cpp [hunk 27]
```diff
@@ -1392,6 +1394,93 @@ namespace model {
...
+        auto plantFuelTypes = p_->heatingFuelTypes();
+        result.insert(plantFuelTypes.begin(), plantFuelTypes.end());
+      }
+      if (auto zoneComp_ = containingZoneHVACComponent()) {
+        // The only accepted types are the WaterHeaterHeatPump or WaterHeaterHeatPumpWrappedCondenser so skip the checks
+        result.insert(FuelType::Electricity);
+      }
+
+      return {result.begin(), result.end()};
+    }
```
Comment (jmarrec) [inline src/model/WaterHeaterMixed.cpp:1465]: cf HPWH

File: src/model/ThermalStorageChilledWaterStratified.cpp (hunk 24 above)
Comment (jmarrec) [reply src/model/ThermalStorageChilledWaterStratified.cpp:754]: Matching develop: https://github.com/NREL/OpenStudio/blob/4fd225f2ce4d7a5bc325337ff580942da38d0017/src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.cpp#L523-L525

File: src/model/ThermalStorageIceDetailed.cpp (hunk 25 above)
Comment (jmarrec) [reply src/model/ThermalStorageIceDetailed.cpp:300]: Matching develop: https://github.com/NREL/OpenStudio/blob/4fd225f2ce4d7a5bc325337ff580942da38d0017/src/energyplus/ForwardTranslator/ForwardTranslatePlantEquipmentOperationSchemes.cpp#L526-L529

**Review by tijcolem — APPROVED** [review]
